rfd/0001-general-design.md

```diff
@@ -0,0 +1,433 @@
+         │ U │     ┌─┤Request 1├─┐      ┌──┤ R1 ├─▶│  T  │                            │  K  │────┬─┼─▶│  addr1 │  │
+         │ R │     │ └─────────┘ │      │  └────┘  │  E  │                            │  E  │    │ │  └────────┘  │
+         │ E │     │             │      │          │  N  │                            │  N  │    │ │  ┌────────┐  │
+         └─┬─┘┌────────┐         ▼      │          │  D  ├────────────────────────────▶  D  │    ├─┼──▶  addr2 │  │
+           │  │        │     ┌──────┬──────┐       └─────┘                            └─────┘    │ │  └────────┘  │
+           └──│ Client │     │ 8090 │ mTLS │                                                     │ │  ┌────────┐  │
+              │        │     └──────┴──────┘       ┌─────┐                            ┌─────┐    └─┼─▶│  addr3 │  │
+              └────────┘         │      │          │  F  │                            │  B  │      │  └────────┘  │
+                   │             │      │          │  R  ├────────────────────────────▶  A  │      │  ┌────────┐  │
+                   │ ┌─────────┐ │      │          │  O  │                            │  C  │    ┌─┼─▶│  addr4 │  │
+                   └─┤Request 2├─┘      │  ┌────┐  │  N  │                            │  K  │────┤ │  └────────┘  │
+                     └─────────┘        └──┤ R2 ├─▶│  T  │      ┌─────────┐           │  E  │    │ │  ┌────────┐  │
+                                           └────┘  │  E  │ ┌────┤ options ├────┐      │  N  │    └─┼─▶│  addrN │  │
+                                                   │  N  │ │┌───┴─────────┴───┐│      │  D  │      │  └────────┘  │
+                                                   │  D  │ ││addr4,addr5,addrN││      └─────┘      │              │
+                                                   └────┬┴─│├─────────────────┤│──┐      ▲         └──────────────┘
+                                                        └──□│   timeoutSec    ││  │      │                         
+                                                           │└─────────────────┘│  └──────┘                         
+                                                           └───────────────────┘                                   
+```
+#### Frontend
+Is the part of the application customer can configure, and it represents Load Balancer
+object options.
+
+Frontend Available in Api Interface, and is entity in Storage Abstraction.
+
+#### Backend
+Is the internal abstraction created with Frontend options when the Connection Manager receives
+the request. Backends have the following functionality:
+- Provide a routine to pipe the traffic
+- Apply routing strategy
+- Manage connection health per pool
+- Provide balancing with available algorithms
+
+###### Balancing Algorithms
+This design limits initial implementation to have the following algorithms:
+1) Round Robin
+   - implementation in a form of atomic counter to the size of the pool of servers
```

Comment:
LB Requirement for [L5](https://github.com/gravitational/careers/blob/main/challenges/systems/challenge-2.md#level-5) doesn't require to implement round robing so please feel to cut the scope.